src/main/java/ch/uzh/soprafs22/groupmatcher/service/MatcherService.java

```diff
@@ -118,4 +124,62 @@ public double[][] buildAnswersMatrixForMatcher(Matcher matcher) {
                     .map(selectedAnswer -> selectedAnswer.getQuestion().getAnswers().size()).orElse(0))
                     .mapToDouble(Integer::doubleValue).toArray()).toArray(double[][]::new);
     }
+
+    public double[][] buildAnswersMatrixForMatcherByQCategory(Matcher matcher, QuestionCategory qCategory){
+
+        List<Long> answersIds = matcher.getQuestions().stream()
+                .filter(question -> question.getQuestionCategory().equals(qCategory))
+                .flatMap(question -> question.getAnswers().stream().map(Answer::getId))
+                .toList();
+
+        return matcher.getStudents().stream().map(student ->
+                answersIds.stream().map(answerId -> student.getSelectedAnswers()
+                    .stream().filter(selectedAnswer -> selectedAnswer.getId().equals(answerId)).findFirst()
+                    .map(selectedAnswer -> selectedAnswer.getQuestion().getAnswers().size()).orElse(0))
+                    .mapToDouble(Integer::doubleValue).toArray()).toArray(double[][]::new);
+    }
+
+    public Matcher runMatchingPrimSpanningTree(Matcher matcher){
```

Comment:
Rather than setting up the whole data processing and graph building ourselves I suggest we use an external library, specificall on called JGraphT.

I would suggest this setup (renaming runMatchingPrimSpanningTree to runBalancedSkillsModel):

```
private double calculateMatchingScoreForPair(Student student1, Student student2) {
        return student1.getSelectedAnswers().stream().mapToDouble(selectedAnswer ->
                student2.getSelectedAnswers().contains(selectedAnswer) ? selectedAnswer.calculateMatchingScore() : 0.0).sum();
    }

public Matcher runBalancedSkillsModel(Matcher matcher) {
        Graph<Student, DefaultWeightedEdge> studentsGraph = new SimpleWeightedGraph<>(DefaultWeightedEdge.class);
        Graphs.addAllVertices(studentsGraph, matcher.getStudents());
        Sets.combinations(studentsGraph.vertexSet(), 2).forEach(studentsPairSet -> {
            List<Student> studentsPair = studentsPairSet.stream().toList();
            Student student1 = studentsPair.get(0);
            Student student2 = studentsPair.get(1);
            double matchingScore = calculateMatchingScoreForPair(student1, student2);
            if (matchingScore != 0)
                studentsGraph.setEdgeWeight(studentsGraph.addEdge(student1, student2), matchingScore);
        });
        PrimMinimumSpanningTree<Student, DefaultWeightedEdge> studentsPrimTree = new PrimMinimumSpanningTree<>(studentsGraph);
        studentsPrimTree.getSpanningTree().forEach(edge -> {
                // TODO
        });
        return matcherRepository.save(matcher);
    }
```

With this we don't need all the modules under matching/ and we don't need the method buildAnswersMatrixForMatcherByQCategory as well. The only thing to still figure out is how to get the data we want from the studentsPrimTree that we get from JGraphT - I suggest checking out the library docs [here](https://jgrapht.org/guide/UserOverview#graph-algorithms) and the package javadocs [here](https://jgrapht.org/javadoc/org.jgrapht.core/org/jgrapht/alg/spanning/PrimMinimumSpanningTree.html).

In this case I personally think it's really worth the hassle to use an external library because that way we don't need to worry about the implementation and we can easily switch the type of algorithm if we want - JGraphT includes 20+ different algorithms like you can see [here](https://jgrapht.org/javadoc/org.jgrapht.core/org/jgrapht/alg/interfaces/package-summary.html)

(Note: to make `selectedAnswer.calculateMatchingScore()` work you need to add this function inside the Question class, see the comment there)

Reply:
Actually the previous score calculation was wrong (since it will only be called when we want BALANCED_SKILLS matching strategy), here is the corrected version:

```
    private double calculateBalancedMatchingScoreForPair(Student student1, Student student2) {
        return student1.getSelectedAnswers().stream().mapToDouble(selectedAnswer ->
                 selectedAnswer.calculateBalancedMatchingScore(student2.getId())).sum();
    }
```
(Note method name was changed from `calculateMatchingScoreForPair` to `calculateBalancedMatchingScoreForPair`)

Then inside the Answer entity add at the end:

```
    public Double calculateBalancedMatchingScore(Long otherStudentId) {
        boolean isSharedAnswer = students.stream().anyMatch(student -> student.getId().equals(otherStudentId));
        return question.getQuestionCategory().calculateBalancedMatchingScore(isSharedAnswer);
    }
```

And the QuestionCategory enum will look like this:

```
@Getter
@AllArgsConstructor
public enum QuestionCategory {
     PREFERENCES(true), SKILLS(false);

     private final boolean similarityMatching;

     public Double calculateBalancedMatchingScore(boolean isSharedAnswer) {
         return isSharedAnswer == similarityMatching ? 1.0 : 0.0;
     }
}
```

Reply:
JGraphT library will be implemented via other development tasks in a proper way.